### PR TITLE
[PORT] Add support for operations on arrays

### LINKED ISSUE #1994: [PORT] Add support for operations on arrays
> Port this change from botbuilder-dotnet/master branch:
https://github.com/microsoft/botbuilder-dotnet/pull/3669

This adds support for operations on arrays.  The API surface area is the same except that the OnClearProperty event was removed.  The internals were also simplified to make it clearer what is going on.  There will be a separate check-in that updates the samples to use arrays.  I wanted to keep this review simpler.

# Changed projects
* Microsoft.Bot.Builder.AI.LUIS
* Microsoft.Bot.Builder.Dialogs.Adaptive
* Microsoft.Bot.Builder.Dialogs
* Microsoft.Bot.Builder.Dialogs.Adaptive.Tests

### PR DESCRIPTION
**Pull request by chon219**
Fixes #1994 

## Description

> Port this change from botbuilder-dotnet/master branch:
> microsoft/botbuilder-dotnet#3669

This adds support for operations on arrays. The API surface area is the same except that the OnClearProperty event was removed. The internals were also simplified to make it clearer what is going on. There will be a separate check-in that updates the samples to use arrays. I wanted to keep this review simpler.

## Specific Changes

  - 
  -
  -

### CODE REVIEW

File: libraries/botbuilder-dialogs/src/memory/dialogPath.ts [hunk 1]
```diff
@@ -13,6 +13,9 @@ export class DialogPath {
     /// Currently expected properties.
     public static readonly expectedProperties = 'dialog.expectedProperties';
 
+    /// Default operation to use for entities where there is no identified operation entity.
+    public static readonly defaultOperation = 'dialog.defaultOperation';
```
Comment (chrimc62) [inline libraries/botbuilder-dialogs/src/memory/dialogPath.ts:17]: >defaultOperation [](start = 54, length = 16)

Does this need to be $defaultOperation?  I ask because that is the key in adaptiveDialog.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialogComponentRegistration.ts [hunk 2]
```diff
@@ -12,7 +12,7 @@ import { AdaptiveTypeBuilder } from './adaptiveTypeBuilder';
...
 import { Ask, AttachmentInput, ChoiceInput, ConfirmInput, DateTimeInput, NumberInput, OAuthInput, TextInput, AttachmentOutputFormat, ChoiceOutputFormat } from './input';
-import { OnActivity, OnAssignEntity, OnBeginDialog, OnCancelDialog, OnChooseEntity, OnChooseIntent, OnChooseProperty, OnClearProperty, OnCondition, OnConversationUpdateActivity, OnDialogEvent, OnEndOfActions, OnEndOfConversationActivity, OnError, OnEventActivity, OnHandoffActivity, OnIntent, OnInvokeActivity, OnMessageActivity, OnMessageDeleteActivity, OnMessageReactionActivity, OnMessageUpdateActivity, OnQnAMatch, OnRepromptDialog, OnTypingActivity, OnUnknownIntent } from './conditions';
```
Comment (chrimc62) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialogComponentRegistration.ts:15]: Why was this removed?

**Comment by chrimc62** [comment]
>     /**

Document

---
Refers to: libraries/botbuilder-dialogs-adaptive/src/entityInfo.ts:12 in 9e072ea. [](commit_id = 9e072eaf5ed7f7c782bc7d9413fcb2c6303aaaaa, deletion_comment = False)

**Comment by chrimc62** [comment]
>     /**

Document

---
Refers to: libraries/botbuilder-dialogs-adaptive/src/schemaHelper.ts:11 in 9e072ea. [](commit_id = 9e072eaf5ed7f7c782bc7d9413fcb2c6303aaaaa, deletion_comment = False)

File: libraries/botbuilder-dialogs/src/memory/dialogPath.ts (hunk 1 above)
Comment (chon219) [reply libraries/botbuilder-dialogs/src/memory/dialogPath.ts:17]: > > defaultOperation [](start = 54, length = 16)
> 
> Does this need to be $defaultOperation? I ask because that is the key in adaptiveDialog.

I don't think so, we have a private property in adaptiveDialog named _defaultOperationKey_ which represents the key $defaultOperation, which is a part of schema.

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignments.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+     * @param actionContext Memory context.
+     */
+    public write(actionContext: ActionContext): void {
+        actionContext.state.setValue(events, this);
+    }
+
+    /**
+     * Returns the next entity event to surface.
+     */
+    public nextAssignment(): Partial<EntityAssignment> {
```
Comment (stevengum) [inline libraries/botbuilder-dialogs-adaptive/src/entityAssignments.ts:44]: This seems like it should just be a getter. And possibly renamed?

Comment (chon219) [reply libraries/botbuilder-dialogs-adaptive/src/entityAssignments.ts:44]: Got it.
